design/design-doc-smaller-providers.md

```diff
@@ -0,0 +1,429 @@
+plane nodes and the version of Kubernetes being run. As recently as March 2023
+we’ve seen that installing just provider-aws on a new GKE cluster can bring the
+control plane offline for an hour while it scales to handle the load.
+
+Problems tend to start happening once you’re over - very approximately - 500
+CRDs. Keep in mind that many folks are installing Kubernetes alongside other
+tools that use CRDs. Here’s a brief survey of the number of CRDs common tools
+install:
+
+* Crossplane (without any providers) - 12
+* Google Config Connector - ~200
+* Azure Service Operator - ~190
+* Kyverno - 12
+* Istio - 14
+* Argo Workflows - 8
+
+Installing too many CRDs also affects the performance of Kubernetes clients like
+kubectl, Helm, and ArgoCD. These clients are often built under the assumption
+that there won’t be more than 50-100 CRDs installed on a cluster.
+
+You can read more about the issues Kubernetes faces when too many CRDs are
+present at:
+
+* https://github.com/crossplane/crossplane/blob/v1.11.2/design/one-pager-crd-scaling.md
+* https://blog.upbound.io/scaling-kubernetes-to-thousands-of-crds/
+
+Apart from performance issues, some folks have expressed concerns about the
+security and general “bloat” of having many unused CRDs installed. I believe the
+security concerns to be an educational issue - I haven't yet seen a convincing
+argument for removing CRDs as opposed to using RBAC.
+
+Crossplane maintainers have invested in improving the performance of the
+Kubernetes control plane and clients in the face of many CRDs. These
+improvements have been insufficient to alleviate community pain. The assumption
+that there won’t be “too many” CRDs is pervasive in the Kubernetes codebase.
+There’s no one easy fix, and the conservative speed at which cloud providers and
+their customers pick up new Kubernetes releases means it can take years for
+fixes to be deployed.
+
+Allowing folks to lower the ratio of installed-to-used CRDs has been proposed
+for some time. The Crossplane maintainers are now aligned that it’s necessary.
+We have explored the following options:
+
+* Make it possible to filter unnecessary CRDs by configuring a provider.
+* Break the biggest providers up into smaller, service-scoped providers.
+* Load CRDs into the API server lazily, on-demand.
+
+## Proposal
+
+I propose that provider maintainers break the largest providers into smaller,
+service-scoped providers. Each service-scoped provider would correspond to what
+is currently an API group within a provider. For example,
+`upbound/provider-aws-rds` would extend Crossplane with support for the 22
+managed resources (and thus 22 CRDs) in the `rds.aws.upbound.io` API group.
+
+### Overview
+
+I proposed we only break up the largest providers. For example:
+
+* `upbound/provider-aws` - Becomes ~150 smaller providers.
+* `upbound/provider-azure` - Becomes ~100 smaller providers.
+* `upbound/provider-gcp` - Becomes ~75 smaller providers.
```

Comment:
It's outside the scope of this design, but for the record we have commitment from folks at Upbound to update:

* Any large official providers we maintain (i.e. AWS, GCP, and Azure).
* The `upjet` tooling.
* The Upbound marketplace